hal/src/i2c.rs

```diff
@@ -0,0 +1,412 @@
+                        // t_SCL ~= t_SYNC1 + t_SYNC2 + t_SCLL + t_SCLH
+                        let i2cclk = Self::clock(rcc).0;
+                        let ratio = i2cclk / freq.integer() - 4;
+                        let (presc, scll, sclh, sdadel, scldel) = if freq >= 100.kHz() {
+                            // fast-mode or fast-mode plus
+                            // here we pick SCLL + 1 = 2 * (SCLH + 1)
+                            let presc = ratio / 387;
+
+                            let sclh = ((ratio / (presc + 1)) - 3) / 3;
+                            let scll = 2 * (sclh + 1) - 1;
+
+                            let (sdadel, scldel) = if freq > 400.kHz() {
+                                // fast-mode plus
+                                let sdadel = 0;
+                                let scldel = i2cclk / 4_000_000 / (presc + 1) - 1;
+
+                                (sdadel, scldel)
+                            } else {
+                                // fast-mode
+                                let sdadel = i2cclk / 8_000_000 / (presc + 1);
+                                let scldel = i2cclk / 2_000_000 / (presc + 1) - 1;
+
+                                (sdadel, scldel)
+                            };
+
+                            (presc, scll, sclh, sdadel, scldel)
+                        } else {
+                            // standard-mode
+                            // here we pick SCLL = SCLH
+                            let presc = ratio / 514;
+
+                            let sclh = ((ratio / (presc + 1)) - 2) / 2;
+                            let scll = sclh;
+
+                            let sdadel = i2cclk / 2_000_000 / (presc + 1);
+                            let scldel = i2cclk / 800_000 / (presc + 1) - 1;
+
+                            (presc, scll, sclh, sdadel, scldel)
+                        };
+
+                        assert!(presc < 16);
+                        assert!(scldel < 16);
+                        assert!(sdadel < 16);
+                        let sclh = u8::try_from(sclh).unwrap();
+                        let scll = u8::try_from(scll).unwrap();
+
+                        // Configure for "fast mode" (400 KHz)
+                        // NOTE(write): writes all non-reserved bits.
+                        i2c.timingr.write(|w| {
+                            w.presc()
+                                .bits(presc as u8)
+                                .sdadel()
+                                .bits(sdadel as u8)
+                                .scldel()
+                                .bits(scldel as u8)
+                                .scll()
+                                .bits(scll)
+                                .sclh()
+                                .bits(sclh)
+                        });
+
+                        // Enable the peripheral
+                        i2c.cr1.modify(|_, w| w.pe().set_bit());
```

Comment:
This should be a write since there are no writes to `cr1` between the reset and this point.